From 7ed592c862936d227aaccd1532c6a3a423528aa5 Mon Sep 17 00:00:00 2001
From: Toni Heidenreich <tonihei@google.com>
Date: Tue, 8 Jul 2025 05:42:23 -0700
Subject: [PATCH] Add flag to exclude capability propagation when connecting
 MediaBrowser

Also assume this flag is true when EXTRA_RECENT is used as querying
the recent playback should always only be a one-off short lived request

Manual cherry-pick of ag/33253388

Bug: 393582077
(cherry picked from https://googleplex-android-review.googlesource.com/q/commit:c61afbfe43c3da7b72408a43b44b03bb13c3243e)
Merged-In: Idc9b428f2ff53cb490ee596da35d24e1e7fca681
Change-Id: Idc9b428f2ff53cb490ee596da35d24e1e7fca681
---
 media/java/android/media/browse/MediaBrowser.java    | 10 ++++++++--
 .../android/service/media/MediaBrowserService.java   | 12 ++++++++++++
 2 files changed, 20 insertions(+), 2 deletions(-)

diff --git a/media/java/android/media/browse/MediaBrowser.java b/media/java/android/media/browse/MediaBrowser.java
index 022278298875..fc2a53a05af7 100644
--- a/media/java/android/media/browse/MediaBrowser.java
+++ b/media/java/android/media/browse/MediaBrowser.java
@@ -37,6 +37,7 @@ import android.os.ResultReceiver;
 import android.service.media.IMediaBrowserService;
 import android.service.media.IMediaBrowserServiceCallbacks;
 import android.service.media.MediaBrowserService;
+import android.service.media.MediaBrowserService.BrowserRoot;
 import android.text.TextUtils;
 import android.util.ArrayMap;
 import android.util.Log;
@@ -183,8 +184,13 @@ public final class MediaBrowser {
 
                 boolean bound = false;
                 try {
-                    bound = mContext.bindService(intent, mServiceConnection,
-                            Context.BIND_AUTO_CREATE | Context.BIND_INCLUDE_CAPABILITIES);
+                    int bindServiceFlags = Context.BIND_AUTO_CREATE;
+                    if (mRootHints == null
+                        || (!mRootHints.containsKey(BrowserRoot.EXTRA_EXCLUDE_CAPABILITIES)
+                                && !mRootHints.containsKey(BrowserRoot.EXTRA_RECENT))) {
+                        bindServiceFlags |= Context.BIND_INCLUDE_CAPABILITIES;
+                    }
+                    bound = mContext.bindService(intent, mServiceConnection, bindServiceFlags);
                 } catch (Exception ex) {
                     Log.e(TAG, "Failed binding to service " + mServiceComponent);
                 }
diff --git a/media/java/android/service/media/MediaBrowserService.java b/media/java/android/service/media/MediaBrowserService.java
index 39ef528fb69b..44e5bbd9f28e 100644
--- a/media/java/android/service/media/MediaBrowserService.java
+++ b/media/java/android/service/media/MediaBrowserService.java
@@ -608,6 +608,18 @@ public abstract class MediaBrowserService extends Service {
          */
         public static final String EXTRA_SUGGESTED = "android.service.media.extra.SUGGESTED";
 
+        /**
+         * A key for the root hint when connecting a browser to indicate that the bind connection
+         * should not include capabilities.
+         *
+         * <p>When using {@link #EXTRA_RECENT}, this flag is assumed to be set as well.
+         *
+         * @see android.content.Context#BIND_INCLUDE_CAPABILITIES
+         * @hide
+         */
+        public static final String EXTRA_EXCLUDE_CAPABILITIES =
+                "android.service.media.extra.EXCLUDE_CAPABILITIES";
+
         private final String mRootId;
         private final Bundle mExtras;
 
-- 
2.52.0.rc1.455.g30608eb744-goog

